File: src/main/java/vip/mystery0/base/springboot/utils/trace/TraceLogUtil.java
package vip.mystery0.base.springboot.utils.trace;

import org.apache.tomcat.util.buf.StringUtils;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.slf4j.MDC;
import vip.mystery0.base.springboot.constant.Constants;
import vip.mystery0.tools.java.factory.JsonFactory;
import vip.mystery0.tools.java.utils.TimeUtil;

import javax.servlet.http.HttpServletRequest;
import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.List;

/**
 * @author mystery0
 */
public class TraceLogUtil {
    private static final Logger log = LoggerFactory.getLogger(TraceLogUtil.class);

    /**
     * 记录请求
     *
     * @param request   请求体
     * @param maxLength 参数打印最大长度
     */
    public static void logRequest(HttpServletRequest request,
                                  int maxLength) {
        List<String> params = new ArrayList<>();
        request.getParameterMap().forEach((s, strings) -> {
            String value = StringUtils.join(strings);
            if (value.length() > maxLength) {
                String start = value.substring(0, 4);
                String end = value.substring(value.length() - 4);
                params.add(s + "=>" + start + "...." + end);
            } else {
                params.add(s + "=>" + value);
            }
        });
        String args = StringUtils.join(params);

        log.info("╔══════════════════════");
        log.info("║ {}", TimeUtil.formatDateTime(LocalDateTime.now()));
        log.info("║ {}", MDC.get(Constants.MDC_URI));
        if (!"".equals(args)) {
            log.info("║ params: 【{}】", params);
        }
        log.info("║ IP: {}", TraceHelper.getClientIP(request));
        log.info("╙──────────────────────");
    }

    public static void logResponse(Object result) {
        log.info("╓──────────────────────");
        Long costTime = getCostTime();
        if (costTime != null) {
            log.info("║ request cost time: {}ms", costTime);
        }
        log.info("║ {}", MDC.get(Constants.MDC_URI));
        if (result != null) {
            String json = JsonFactory.toJson(result);
            log.info("║ return: {}", json);
        }
        log.info("╚══════════════════════");
    }

    private static Long getCostTime() {
        try {
            return (System.currentTimeMillis() - Long.parseLong(MDC.get(Constants.MDC_START_TIME)));
        } catch (Exception e) {
            log.warn("get cost time failed");
            return null;
        }
    }
}

File: src/main/java/vip/mystery0/base/springboot/utils/trace/TraceLogUtil.kt
package vip.mystery0.base.springboot.utils.trace

import org.aspectj.lang.ProceedingJoinPoint
import org.slf4j.LoggerFactory
import org.slf4j.MDC
import vip.mystery0.base.springboot.constant.MDC_START_TIME
import vip.mystery0.base.springboot.constant.MDC_URI
import vip.mystery0.tools.kotlin.factory.toJson
import vip.mystery0.tools.kotlin.utils.doNoException
import vip.mystery0.tools.kotlin.utils.formatDateTime
import java.time.LocalDateTime
import javax.servlet.http.HttpServletRequest

/**
 * @author mystery0
 */
object TraceLogUtil {
    private val log = LoggerFactory.getLogger(TraceLogUtil::class.java)

    fun <R : Any?> Pair<R?, Exception?>.log(): R? {
        if (second != null)
            log.error("do something error", second)
        return first
    }

    /**
     * 记录请求
     *
     * @param request   请求体
     */
    fun logRequest(request: HttpServletRequest, joinPoint: ProceedingJoinPoint) {
        val args = doNoException { joinPoint.args.toJson() }
        log.info("╔══════════════════════")
        log.info("║ {}", LocalDateTime.now().formatDateTime())
        log.info("║ {} {}", request.method, request.requestURI)
        if (!args.isNullOrBlank())
            log.info("║ params: 【{}】", args)
        log.info("║ IP: {}", TraceHelper.getClientIP(request))
        log.info("╙──────────────────────")
    }

    fun logRequestBody(request: HttpServletRequest) {
        val body = doNoException { String(request.inputStream.readBytes()) }
        log.info("╔══════════════════════")
        log.info("║ {}", LocalDateTime.now().formatDateTime())
        log.info("║ {} {}", request.method, request.requestURI)
        if (!body.isNullOrBlank())
            log.info("║ params: 【{}】", body)
        log.info("║ IP: {}", TraceHelper.getClientIP(request))
        log.info("╙──────────────────────")
    }

    fun logResponse(result: Any?) {
        log.info("╓──────────────────────")
        if (costTime != null)
            log.info("║ request cost time: {}ms", costTime)
        log.info("║ {}", MDC.get(MDC_URI))
        if (result != null)
            log.info("║ return: {}", result.toJson())
        log.info("╚══════════════════════")
    }

    private val costTime: Long?
        get() = try {
            if (MDC.get(MDC_START_TIME) == null) null else System.currentTimeMillis() - MDC.get(MDC_START_TIME).toLong()
        } catch (e: Exception) {
            log.warn("get cost time failed")
            null
        }
}